Add function to materialize COW storages

kurtamohler opened the pull request:
Part of #109833

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113397
* __->__ #113396



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

Differential Revision: [D51466282](https://our.internmc.facebook.com/intern/diff/D51466282)

kurtamohler committed:
Add function to materialize COW storages

[ghstack-poisoned]

kurtamohler committed:
Update on "Add function to materialize COW storages"




[ghstack-poisoned]

kurtamohler committed:
Update on "Add function to materialize COW storages"




[ghstack-poisoned]

kurtamohler committed:
Update on "Add function to materialize COW storages"




[ghstack-poisoned]

kurtamohler committed:
Update on "Add function to materialize COW storages"




[ghstack-poisoned]

kurtamohler committed:
Update on "Add function to materialize COW storages"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kurtamohler committed:
Update on "Add function to materialize COW storages"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kurtamohler commented on c10/core/impl/cow/COW.cpp:
```diff
@@ -30,24 +29,18 @@ at::DataPtr copy_data_ptr(at::DataPtr const& data_ptr) {
-    const void* context,
-    const void* data,
-    const at::Allocator* allocator) {
-  if (allocator == c10::GetDefaultMobileCPUAllocator()) {
```
I realized that I could remove the dependence on CPUAllocator here by adding a new method `Allocator::is_simple_data_ptr`. It's better for CPUAllocator to override that method than to have a conditional here in `COW.cpp`

Also, since `StorageImpl` now needs to have access to `materialize_cow_storage`, I needed to combine the `impl_cow` bazel library back into the `base` library

kurtamohler committed:
Update on "Add function to materialize COW storages"

Part of #109833




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kurtamohler committed:
Update on "Add function to materialize COW storages"

Part of #109833




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/Resize.cpp:
```diff
@@ -94,13 +94,14 @@ void resize_bytes_cpu(StorageImpl* storage, size_t size_bytes) {
+    memcpy(new_data.get(), old_data.get(), copy_capacity);
   }
+  storage->set_data_ptr_noswap(std::move(new_data));
+  storage->set_nbytes(size_bytes);
```
The new code looks right to me, but how come you needed to do this?

kurtamohler replied:
Since `set_data_ptr` returns a mutable DataPtr, it causes a materialization. So I rewrote this to avoid the unnecessary materialization. `set_data_ptr_noswap` doesn't materialize the old DataPtr

ezyang commented:
@pytorchbot merge

DanilBaibak commented:
@pytorchbot revert -m "Break internal build" -c ghfirst

DanilBaibak commented:
@kurtamohler sorry, I have to revert your PR. It leads to internal failures. The build failure is:
```
fbcode/caffe2/caffe2/fb/init/huge_pages_allocator.cc:171:8: error: abstract class is marked 'final' [-Werror,-Wabstract-final-class]
[CONTEXT] struct OversizeArenaAllocator final : at::Allocator {
[CONTEXT]        ^
fbcode/caffe2/caffe2/fb/init/huge_pages_allocator.cc:208:31: error: variable type 'caffe2::(anonymous namespace)::OversizeArenaAllocator' is an abstract class
[CONTEXT] static OversizeArenaAllocator g_oversize_cpu_alloc;
[CONTEXT]                               ^
```

Using `virtual void copy_data(void* dest, const void* src, std::size_t count) const = 0;` is harder to roll out without updating the usage of Allocator class internally. @ezyang, could you please help with this?

ezyang commented:
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51466282).

kurtamohler committed:
Update on "Add function to materialize COW storages"

Part of #109833




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51466282](https://our.internmc.facebook.com/intern/diff/D51466282)

[ghstack-poisoned]

kurtamohler commented:
@ezyang, the ASAN error should be fixed now, though I wasn't able to get an ASAN build working in my environment to verify for sure. The problem was that `testing::StrEq` is a bad way to test that storages have the same data, since storages are not null-terminated

ezyang commented:
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51466282).

ezyang commented:
Hey Kurt, I'm still running into weird internal fbcode build problems from having the cow/ subdirectory. I could probably, in principle, figure it out, but it seems a lot simpler to just get rid of the cow/ folder. Do you think you could do that?

kurtamohler commented:
Sure thing, I'll do that

ezyang commented:
Also, I'm traveling to China until the new year, so you'll need someone else to help out with the fbcode side. Maybe @malfet is willing to help

kurtamohler committed:
Update on "Add function to materialize COW storages"

Part of #109833




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51466282](https://our.internmc.facebook.com/intern/diff/D51466282)

[ghstack-poisoned]

kurtamohler commented:
I've removed the `c10/core/impl/cow` directory and moved its contents to `c10/core/impl`, so hopefully those internal errors are gone now

kurtamohler committed:
Update on "Add function to materialize COW storages"

Part of #109833




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51466282](https://our.internmc.facebook.com/intern/diff/D51466282)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51466282).

kurtamohler committed:
Update on "Add function to materialize COW storages"

Part of #109833




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51466282](https://our.internmc.facebook.com/intern/diff/D51466282)

[ghstack-poisoned]

kurtamohler commented:
Hi @malfet, I was wondering if you might have time to take a look at this PR and let me know if the internal errors are fixed or not?

kurtamohler committed:
Update on "Add function to materialize COW storages"

Part of #109833




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51466282](https://our.internmc.facebook.com/intern/diff/D51466282)

[ghstack-poisoned]

kurtamohler commented:
Hey @ezyang, could you let me know if the internal errors are fixed?

ezyang commented:
I'm baaack

ezyang commented:
Testing on https://www.internalfb.com/diff/D52610522 because I can no longer ghimport this PR smh

kurtamohler commented:
Closing this since it was moved to #117053
